app/controllers/application_controller/tags.rb

```diff
@@ -185,6 +163,8 @@ def tagging_edit_tags_save_and_replace_right_cell
 
   # Add/remove tags in a single transaction
   def tagging_save_tags
+    new_assignments = JSON.parse(params['data']).map { |tag| tag['values'].map { |v| v['id'] } }
+    @edit[:new][:assignments] = new_assignments.flatten
```

Comment:
What about using `flat_map`? Generally I don't like the recursive `map`s but if it's not possible to `reduce` their number (pun intended), at least get rid of the `flatten`.

As far as I understand this, you're in charge of both the client side data generation for the request and the server side data parsing, so worst case you can come up with a simpler format that doesn't need this much preprocessing.